pkg/controller/operators/internal/alongside/alongside.go

```diff
@@ -0,0 +1,83 @@
+// Package alongside provides a mechanism for recording the fact that
+// one object was installed alongside another object as part of the
+// installation of the same operator version.
+package alongside
+
+import (
+	"fmt"
+	"hash/fnv"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	prefix = "operatorframework.io/installed-alongside-"
+)
+
+// NamespacedName is a reference to an object by namespace and name.
+type NamespacedName struct {
+	Namespace string
+	Name      string
+}
+
+// Annotator translates installed-alongside references to and from
+// object annotations.
+type Annotator struct{}
+
+// FromObject returns a slice containing each namespaced name
+// referenced by an alongside annotation on the provided Object.
+func (a Annotator) FromObject(o metav1.Object) []NamespacedName {
+	var result []NamespacedName
+	for k, v := range o.GetAnnotations() {
+		if !strings.HasPrefix(k, prefix) {
+			continue
+		}
+		tokens := strings.Split(v, "/")
+		if len(tokens) != 2 {
+			continue
+		}
+		result = append(result, NamespacedName{
+			Namespace: tokens[0],
+			Name:      tokens[1],
+		})
+	}
+	return result
+}
+
+// ToObject removes all existing alongside annotations on the provided
+// Object and adds one new annotation per entry in the provided slice
+// of namespaced names.
+func (a Annotator) ToObject(o metav1.Object, nns []NamespacedName) {
+	annotations := o.GetAnnotations()
+
+	for key := range annotations {
+		if strings.HasPrefix(key, prefix) {
+			delete(annotations, key)
+		}
+	}
+
+	if len(nns) == 0 {
+		if len(annotations) == 0 {
```

Comment:
nit: is this `if` required? the zero value of a map is nil (which always has length zero)

Reply:
At this point `annotations` could either be `nil` or the zero value. I think the point is to ensure we always set the annotations on `o` to `nil` in this situation.